### PR TITLE
Record username and user_id for all clusters

### PR DESCRIPTION
**Pull request by JohnGarbutt**
Make the API show who created any updated both
CaaS clusters and K8s clusters.

This depends on the CRD changes here:
https://github.com/stackhpc/azimuth-caas-operator/pull/89
https://github.com/stackhpc/azimuth-capi-operator/pull/105

### CODE REVIEW

**Comment by JohnGarbutt** [comment]
Note, you can try these on my dev setup here:
http://portal.apps.128-232-226-118.sslip.io/

File: api/azimuth/cluster_api/dto.py [hunk 1]
```diff
@@ -133,6 +133,12 @@ class Cluster:
     services: t.List[Service]
     #: The time at which the cluster was created
     created_at: datetime.datetime
+    #: Details about the users interacting with the cluster
+    created_by_username: t.Optional[str]
+    created_by_user_id: t.Optional[str]
+    updated_by_username: t.Optional[str]
+    updated_by_user_id: t.Optional[str]
+    is_mine: bool
```
Comment (JohnGarbutt) [inline api/azimuth/cluster_api/dto.py:141]: I am not sure I like its, maybe created_by_current_user is better?

File: api/azimuth/cluster_api/base.py [hunk 2]
```diff
@@ -196,7 +196,7 @@ def find_cluster_template(self, id: str) -> dto.ClusterTemplate:
                 .fetch(id)
         )
 
-    def _from_api_cluster(self, cluster, sizes):
+    def _from_api_cluster(self, cluster, sizes, user_id):
```
Comment (mkjpryor) [inline api/azimuth/cluster_api/base.py:199]: `user_id` is passed in but not used

File: api/azimuth/cluster_api/base.py [hunk 3]
```diff
@@ -317,7 +321,8 @@ def clusters(self) -> t.Iterable[dto.Cluster]:
         self._log("Found %s clusters", len(clusters))
         if clusters:
             sizes = list(self._cloud_session.sizes())
-            return tuple(self._from_api_cluster(c, sizes) for c in clusters)
+            user_id = self._cloud_session.user_id()
+            return tuple(self._from_api_cluster(c, sizes, user_id) for c in clusters)
```
Comment (mkjpryor) [inline api/azimuth/cluster_api/base.py:325]: Passing of `user_id` is not required (see above)

File: api/azimuth/cluster_api/base.py [hunk 4]
```diff
@@ -334,7 +339,8 @@ def find_cluster(self, id: str) -> dto.Cluster:
                 .fetch(id)
         )
         sizes = list(self._cloud_session.sizes())
-        return self._from_api_cluster(cluster, sizes)
+        user_id = self._cloud_session.user_id()
+        return self._from_api_cluster(cluster, sizes, user_id)
```
Comment (mkjpryor) [inline api/azimuth/cluster_api/base.py:343]: Same

File: api/azimuth/cluster_api/base.py [hunk 5]
```diff
@@ -466,22 +474,26 @@ def create_cluster(
         })
         # Use the sizes that we already have
         sizes = [control_plane_size] + [ng["machine_size"] for ng in node_groups]
-        return self._from_api_cluster(cluster, sizes)
+        user_id = self._cloud_session.user_id()
+        return self._from_api_cluster(cluster, sizes, user_id)
```
Comment (mkjpryor) [inline api/azimuth/cluster_api/base.py:478]: Same

File: api/azimuth/cluster_api/base.py [hunk 6]
```diff
@@ -466,22 +474,26 @@ def create_cluster(
...
         cluster = (
             self._client
                 .api(AZIMUTH_API_VERSION)
                 .resource("clusters")
                 .patch(cluster, { "spec": spec })
         )
         sizes = list(self._cloud_session.sizes())
-        return self._from_api_cluster(cluster, sizes)
+        user_id = self._cloud_session.user_id()
+        return self._from_api_cluster(cluster, sizes, user_id)
```
Comment (mkjpryor) [inline api/azimuth/cluster_api/base.py:496]: Same

File: api/azimuth/cluster_api/base.py [hunk 7]
```diff
@@ -496,18 +508,18 @@ def upgrade_cluster(
...
-                "spec": {
-                    "templateName": template.id,
-                },
-            }
-        )
+        cluster = ekclusters.patch(cluster, {"spec": spec})
         sizes = list(self._cloud_session.sizes())
-        return self._from_api_cluster(cluster, sizes)
+        user_id = self._cloud_session.user_id()
+        return self._from_api_cluster(cluster, sizes, user_id)
```
Comment (mkjpryor) [inline api/azimuth/cluster_api/base.py:522]: Same

File: api/azimuth/cluster_engine/drivers/crd/driver.py [hunk 8]
```diff
@@ -68,7 +68,7 @@ def get_cluster_types(client) -> t.Iterable[dto.ClusterType]:
     return cluster_types
 
 
-def get_cluster_dto(raw_cluster, status_if_ready: dto.ClusterStatus = None):
+def get_cluster_dto(raw_cluster, user_id, status_if_ready: dto.ClusterStatus = None):
```
Comment (mkjpryor) [inline api/azimuth/cluster_engine/drivers/crd/driver.py:71]: `user_id` is passed but not used

File: api/azimuth/cluster_engine/drivers/crd/driver.py [hunk 9]
```diff
@@ -119,14 +119,18 @@ def get_cluster_dto(raw_cluster, status_if_ready: dto.ClusterStatus = None):
...
         services=[],
+        created_by_username=raw_cluster.spec.get("createdByUsername"),
+        created_by_user_id=raw_cluster.spec.get("createdByUserId"),
+        updated_by_username=raw_cluster.spec.get("updatedByUsername"),
+        updated_by_user_id=raw_cluster.spec.get("updatedByUserId"),
     )
 
 
-def get_clusters(client) -> t.Iterable[dto.Cluster]:
+def get_clusters(client, ctx: dto.Context) -> t.Iterable[dto.Cluster]:
```
Comment (mkjpryor) [inline api/azimuth/cluster_engine/drivers/crd/driver.py:129]: `ctx` is passed but not required

File: api/azimuth/cluster_engine/drivers/crd/driver.py [hunk 10]
```diff
@@ -119,14 +119,18 @@ def get_cluster_dto(raw_cluster, status_if_ready: dto.ClusterStatus = None):
...
     raw_clusters = list(client.api(CAAS_API_VERSION).resource("clusters").list())
     clusters = []
     for raw_cluster in raw_clusters:
-        cluster = get_cluster_dto(raw_cluster)
+        cluster = get_cluster_dto(raw_cluster, ctx.user_id)
```
Comment (mkjpryor) [inline api/azimuth/cluster_engine/drivers/crd/driver.py:133]: Passing of user ID is not required

File: api/azimuth/cluster_engine/drivers/crd/driver.py [hunk 11]
```diff
@@ -182,10 +188,10 @@ def create_cluster(
             "spec": cluster_spec,
         }
     )
-    return get_cluster_dto(cluster)
+    return get_cluster_dto(cluster, ctx.user_id)
```
Comment (mkjpryor) [inline api/azimuth/cluster_engine/drivers/crd/driver.py:191]: Same

File: api/azimuth/cluster_engine/drivers/crd/driver.py [hunk 12]
```diff
@@ -196,10 +202,10 @@ def delete_cluster(client, name: str):
     # returning the ready state will confuse people
     time.sleep(0.1)
     raw_cluster = cluster_resource.fetch(safe_name)
-    return get_cluster_dto(raw_cluster, status_if_ready=dto.ClusterStatus.DELETING)
+    return get_cluster_dto(raw_cluster,  ctx.user_id, status_if_ready=dto.ClusterStatus.DELETING)
```
Comment (mkjpryor) [inline api/azimuth/cluster_engine/drivers/crd/driver.py:205]: Same

File: api/azimuth/cluster_engine/drivers/crd/driver.py [hunk 13]
```diff
@@ -237,7 +248,7 @@ def update_cluster(client, name: str, params: t.Mapping[str, t.Any], version=Non
     # returning the ready state will confuse people
     time.sleep(0.1)
     raw_cluster = cluster_resource.fetch(safe_name)
-    return get_cluster_dto(raw_cluster, status_if_ready=dto.ClusterStatus.CONFIGURING)
+    return get_cluster_dto(raw_cluster,  ctx.user_id, status_if_ready=dto.ClusterStatus.CONFIGURING)
```
Comment (mkjpryor) [inline api/azimuth/cluster_engine/drivers/crd/driver.py:251]: Same

**Review by mkjpryor — CHANGES_REQUESTED** [review]
Just a few unused variables and propagation thereof to remove.

File: api/azimuth/cluster_api/base.py (hunk 2 above)
Comment (JohnGarbutt) [reply api/azimuth/cluster_api/base.py:199]: oops, yes 😵‍💫

File: ui/src/components/pages/tenancy/platforms/clusters/card.js [hunk 14]
```diff
@@ -165,10 +165,18 @@ const ClusterStatusCard = ({ cluster }) => (
                     <th>Created</th>
                     <td>{cluster.created.toRelative()}</td>
                 </tr>
+                <tr>
+                    <th>Created by</th>
+                    <td>{cluster.created_by_username ? cluster.created_by_username : '-'}</td>
```
Comment (mkjpryor) [inline ui/src/components/pages/tenancy/platforms/clusters/card.js:170]: `cluster.created_by_username || '-'` is sufficient here and avoids repetition. Similarly elsewhere.

Comment (mkjpryor) [reply ui/src/components/pages/tenancy/platforms/clusters/card.js:170]: Also, do we want `-` and not `unknown` or similar?

File: ui/src/components/pages/tenancy/platforms/clusters/card.js [hunk 15]
```diff
@@ -423,7 +430,8 @@ export const ClusterCard = ({
                 </Card.Body>
             )}
             <Card.Body className="small text-muted">
-                Updated {updatedAt.toRelative()}
+                Created {cluster.created.toRelative()}<br/>
+                Created by {cluster.created_by_username ? cluster.created_by_username : '-'}
```
Comment (mkjpryor) [inline ui/src/components/pages/tenancy/platforms/clusters/card.js:434]: Personally, I prefer "Created X mins ago by <user>". Do you have a strong preference for this format?

**Review by mkjpryor — CHANGES_REQUESTED** [review]

File: ui/src/components/pages/tenancy/platforms/clusters/card.js (hunk 15 above)
Comment (JohnGarbutt) [reply ui/src/components/pages/tenancy/platforms/clusters/card.js:434]: it looked odd, not sure by the two lines either.

File: ui/src/components/pages/tenancy/platforms/clusters/card.js (hunk 14 above)
Comment (JohnGarbutt) [reply ui/src/components/pages/tenancy/platforms/clusters/card.js]: done.

**Review by mkjpryor — APPROVED** [review]
LGTM

**Comment by JohnGarbutt** [comment]
Needs final tests.

**Pull request merged**